tidy3d/components/types.py

```diff
@@ -182,6 +182,18 @@ def __modify_schema__(cls, field_schema):
 FreqBoundMin = float
 FreqBound = Tuple[FreqBoundMin, FreqBoundMax]
 
+ComplexTensor = Tuple[
```

Comment:
So there are now `ArrayLike` types that currently allow users to supply numpy arrays and have them validated (if desired) for `ndim` or `dtype`.

These arrays are stored internally as `np.ndarray`, but written to json using `.tolist()`

 I wonder if it makes sense to extend some of that to add eg `shape` and then define these tensors as a `constrained_array(dtype=float, shape=(3,3))` and just use the same machinery. Then it could simplify some of the internals eg in our validators.

See [here](https://github.com/flexcompute/tidy3d/blob/develop/tidy3d/components/types.py#L41-L99) for details.

Any thoughts on this? If we do this, it should be reasonably simple to extend using the existing ArrayLike as an example, but I can help out if you need. 

Note, if we do this, it might be good to rebase first to ensure that you have the most recent changes to ArrayLike introduced in either 2.1.0 or 2.0.3 (I forget exactly when).

Reply:
did this, thanks!